### PR TITLE
Generalize `walk!`

### LINKED ISSUE #654: `walk!` can be made generic and space agnostic with a simple normalizing function
If we define a simple function `normalize_position(pos, space)` that takes in a position, and normalizes it so that it is appropriate for the space's extents, then the walk function becomes the trivial:
```julia
function walk!(agent, direction, model)
    newpos = agent.pos .+ direction
    newpos = normalize_position(newpos, model.space)
    move_agent!(agent, newpos, model)
end
```

and works for any space that defines `normalize_position`. Furthermore, it appears to me that we can reduce source code duplications by defining this `normalize_position` function. We should also make it part of the public api.

### PR DESCRIPTION
**Pull request by AayushSabharwal**
Fixes #654 

- [x] Add tests
- [x] Update docs

### CODE REVIEW

File: src/spaces/utilities.jl [hunk 1]
```diff
@@ -153,6 +153,28 @@ end
 #######################################################################################
 # %% Walking
 #######################################################################################
+"""
+    normalize_position(pos, space::Union{AbstractGridSpace,ContinuousSpace})
```
Comment (Datseris) [inline src/spaces/utilities.jl:157]: Given that in general we do not let users handle and access spaces directly, I think the top level and exported version of htis should be with `ABM` instead.

**Comment by AayushSabharwal** [comment]
```julia
julia> a = (1.3, 2.5)
julia> b = (1.2, 2.3)
julia> f(a, b::NTuple{N,T}) where {N,T} = map(t -> clamp(t[1], zero(T), t[2]), zip(a,b))
julia> f(a,b)
2-element Vector{Float64}:
 1.2
 2.3
```

**Comment by AayushSabharwal** [comment]
Conveniently, the tests for `walk!` should just cover `normalize_position`.

**Comment by AayushSabharwal** [comment]
Wait a minute. Why does `spacesize` need to exist if `Base.size` is already overloaded?

**Comment by AayushSabharwal** [comment]
So there's a change in behaviour of `walk!`: Before, in a periodic continuous space of size `(10.0, 10.0)`, moving to `(5.0, 10.0)` clamped to `(5.0, prevfloat(10.0))` but now it wraps to `(5.0, 0.0)`. Personally, it seems like the latter makes more sense: the "last" value in that dimension is `prevfloat(10.0)` after which it should wrap, similar to how `GridSpace` would.

**Comment by AayushSabharwal** [comment]
I ran a quick set of benchmarks:
```julia
julia> using StaticArrays
julia> x = @SVector[1.3, 2.5]
julia> y = @SVector[1.2, 2.3]
julia> a = Tuple(x)
julia> b = Tuple(y)
julia> f1(x, y) = clamp.(x, zero(y), y)
julia> f2(x,y::NTuple{N,T}) where {N,T} = Tuple(clamp.(x, (zero(T) for _ in 1:N), y))
julia> f3(x,y::NTuple{N,T}) where {N,T} = Tuple(map(t -> clamp(t[1], zero(T), t[2]), zip(x,y)))

julia> @benchmark f1($x,$y)
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  2.639 ns … 8.874 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     2.652 ns             ┊ GC (median):    0.00%
 Time  (mean ± σ):   2.684 ns ± 0.202 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █▆▁                                                       ▁
  ███▃▅▆▃▄▄▄▃▁▃▁▃▄▃▃▄▄▁▁▃▃▁▃▁▄▁▁▁▄▁▄▁▁▃▁▃▅▅▁▄▅▆▅▅▆▆▆▇▆▆▇▇▆▆ █
  2.64 ns     Histogram: log(frequency) by time     3.63 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark f2($a,$b)
BenchmarkTools.Trial: 10000 samples with 269 evaluations.
 Range (min … max):  290.405 ns …  14.432 μs  ┊ GC (min … max): 0.00% … 96.26%
 Time  (median):     294.770 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   321.498 ns ± 468.748 ns  ┊ GC (mean ± σ):  5.20% ±  3.49%

  ▇█▄▁▁▁                ▁▁▁                                     ▁
  ████████▇█▇▇▇▅▅▆▅▆▅▇▆████▇▆▆▇▆▆▃▅▄▅▄▃▃▃▄▅▂▃▃▄▃▅▃▃▄▃▄▄▅▄▄▂▂▃▃▂ █
  290 ns        Histogram: log(frequency) by time        508 ns <

 Memory estimate: 224 bytes, allocs estimate: 5.

julia> @benchmark f3($a,$b)
BenchmarkTools.Trial: 10000 samples with 383 evaluations.
 Range (min … max):  249.371 ns …  11.116 μs  ┊ GC (min … max): 0.00% … 96.79%
 Time  (median):     252.815 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   270.655 ns ± 343.287 ns  ┊ GC (mean ± σ):  4.35% ±  3.35%

  ▄██▅▃▂▁▁▁▁▁▂▁▁                ▁ ▁                             ▂
  ████████████████████▇▆▇▆▇▆▇▇▇█████▇▆█████▆▆▅▆▆▆▇▆▆▄▅▄▄▁▃▄▄▃▅▅ █
  249 ns        Histogram: log(frequency) by time        345 ns <

 Memory estimate: 144 bytes, allocs estimate: 4.
```

Is there something I'm missing with `map` that makes it return a `Vector`?

Also, I've always preferred positions to be `StaticArrays` and not `NTuple`, especially considering basically everything we do with them bar indexing arrays treats positions as mathematical vectors. So far, I hadn't come across a reason to suggest the change. As the saying goes if it isn't broke(n), don't fix it. While actually changing this isn't the goal of this PR, I thought I'd bring it up just in case we actually can't beat that performance.

**Comment by Datseris** [comment]
...? 

```
julia> f(a, b::NTuple{N,T}) where {N,T} = clamp.(a, zero(T), b)
f (generic function with 1 method)

julia> f(a,b)
(1.2, 2.3)

julia> using BenchmarkTools

julia> @btime f($a, $b);
  1.500 ns (0 allocations: 0 bytes)
```

**Comment by AayushSabharwal** [comment]
...
All the lengths I went to, to not think of this.
Be right back, Ima go punch a wall.

**Comment by Datseris** [comment]
> Why does spacesize need to exist if Base.size is already overloaded?

I've thought about that, and concluded that `Base.size(continuous space)` doesn't make sense. `Base.size` is about array/collection sizes, while that's not what we want for continuous space. For discrete space it just so happens that "space eize" and "array size" coincide.

**Comment by Datseris** [comment]
To clarify: `map` would always work, as it is _always_ more stable than broadcasting. You just need to map each element rather than zipping:
```
julia> f(a, b::NTuple{N,T}) where {N,T} = map((x, y) -> clamp(x, zero(T), y), a, b)
f (generic function with 1 method)

julia> @btime f($a, $b);
  1.500 ns (0 allocations: 0 bytes)
```
(remember multi-argument form of `map`)

**Comment by AayushSabharwal** [comment]
I had absolutely no idea `map` worked that way. Learn something new every day, I guess.

**Comment by Datseris** [comment]
> Before, in a periodic continuous space of size (10.0, 10.0), moving to (5.0, 10.0) clamped to (5.0, prevfloat(10.0)) but now it wraps to (5.0, 0.0). 

I agree with this change, but let me see the code as well for certainty. Will review later in more detail.

File: docs/src/api.md [hunk 2]
```diff
@@ -79,6 +79,12 @@ genocide!
 sample!
 ```
 
+## Space utility functions
```
Comment (Datseris) [inline docs/src/api.md:82]: Putting these functions here is okay with me.

File: src/spaces/utilities.jl [hunk 3]
```diff
@@ -153,6 +153,30 @@ end
 #######################################################################################
 # %% Walking
 #######################################################################################
+"""
+    normalize_position(pos, model::ABM{<:Union{AbstractGridSpace,ContinuousSpace}})
+
+Return the position `pos` normalized for the extents of the space of the given `model`.
```
Comment (Datseris) [inline src/spaces/utilities.jl:159]: I think here we should add more info, like "for periodic spaces this wraps the position along each dimension, while for non-periodic spaces this clamps the position to the space extent."

File: src/spaces/utilities.jl [hunk 4]
```diff
@@ -153,6 +153,30 @@ end
...
+"""
+normalize_position(pos, model::ABM) = normalize_position(pos, model.space)
+
+function normalize_position(pos, space::ContinuousSpace{D,true}) where {D}
+    return mod.(pos, spacesize(space))
```
Comment (Datseris) [inline src/spaces/utilities.jl:164]: I was using the `mod1` function, and here you use `mod`, what's the change for?

File: src/spaces/utilities.jl [hunk 5]
```diff
@@ -176,46 +200,21 @@ Example usage in [Battle Royale](
...
-    target = min.(target, prevfloat.(spacesize(model)))
-    move_agent!(agent, target, model)
-end
 function walk!(
     agent::AbstractAgent,
     direction::NTuple{D,Float64},
-    model::ABM{<:ContinuousSpace{D,false}}
+    model::ABM{<:ContinuousSpace}
 ) where {D}
-    target = min.(max.(agent.pos .+ direction, 0.0), prevfloat.(spacesize(model)))
```
Comment (Datseris) [inline src/spaces/utilities.jl:218]: Okay, so instead of the `min, max` approach, you are using `clamp`. Have you tested which version is faster?

**Review by Datseris — COMMENTED** [review]
Just for clarity, can you please summarize which versions you swapped with which? It appears at the moment that all previous versions have been completely replaced (no min max, no mod1, but clamp and mod)

File: src/spaces/utilities.jl (hunk 4 above)
Comment (AayushSabharwal) [reply src/spaces/utilities.jl:164]: `mod1(0.0, 2.0) == 2.0`, but continuous space doesn't include it's upper bound, and does include 0

File: src/spaces/utilities.jl (hunk 5 above)
Comment (AayushSabharwal) [reply src/spaces/utilities.jl:218]: Clamp, by almost 2x

**Comment by AayushSabharwal** [comment]
> Just for clarity, can you please summarize which versions you swapped with which? It appears at the moment that all previous versions have been completely replaced (no min max, no mod1, but clamp and mod)

min max is slower, so I used clamp. mod1 is for Grid Space, since it is basically like mod for 1-indexed arrays. mod is for continuous space, since it wraps around like you'd expect. I just realised there's no need for `prevfloat`, I'll fix that.

File: src/spaces/utilities.jl (hunk 5 above)
Comment (AayushSabharwal) [reply src/spaces/utilities.jl:218]:
```julia
julia> a = (1.3, 2.5)
julia> b = (1.2, 2.3)
julia> f1(a,b) = clamp.(a, 0., b)
julia> f2(a,b) = min.(max.(a, 0.), b)

julia> @benchmark f1($a, $b)
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  1.557 ns … 6.759 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     1.566 ns             ┊ GC (median):    0.00%
 Time  (mean ± σ):   1.591 ns ± 0.136 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █▇▃  ▁                                                    ▂
  ███▅▃█▇▃▆▄▄▁▃▃▃▃▁▁▁▁▁▁▃▃▁▁▁▁▁▁▁▁▁▄▅▄▅▇▆▇▇▇▇▆▇▇▇▇▇▇▇▆▇██▇▇ █
  1.56 ns     Histogram: log(frequency) by time     2.07 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark f2($a, $b)
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  2.279 ns … 42.178 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     2.292 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   2.454 ns ±  1.387 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █▅▄                                                        ▁
  ████▆▄▅▄▃▁▃▃▃▃▃▁▄▄▃▃▃▁▁▄▅▃▄▅▄▅▅▅▆▆▅▆▅▅▆▆▆▆▆▆▅▆▆▆▆▆▆▆▆▆▅▆▆▆ █
  2.28 ns      Histogram: log(frequency) by time     4.39 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

**Review by Datseris — APPROVED** [review]
Does `spacesize` have a documentation string...?

**Comment by AayushSabharwal** [comment]
I moved `spacesize(model::ABM)` to `utilities.jl` since it didn't seem to make sense having it in `continuous.jl`, and added the docstring.

File: src/spaces/utilities.jl [hunk 6]
```diff
@@ -232,3 +232,11 @@ walk!(agent, ::typeof(rand), model::ABM{<:AbstractGridSpace{D}}; kwargs...) wher
 
 walk!(agent, ::typeof(rand), model::ABM{<:ContinuousSpace{D}}) where {D} =
     walk!(agent, Tuple(2.0 * rand(model.rng) - 1.0 for _ in 1:D), model)
+
+"""
+    spacesize(model::ABM)
+
+Return the size of the model's space. Works for [`AbstractGridSpace`](@ref) and
+[`ContinuousSpace`](@ref).
```
Comment (Datseris) [inline src/spaces/utilities.jl:240]: Is there an easy way to make this also work for OSM space, which would return an ecnlosing rectangle with min max lon and min max lat?

**Comment by Datseris** [comment]
We can possibly extent `spacesize` to e.g., `OSMSpace` in a different PR.

**Comment by Datseris** [comment]
@AayushSabharwal as this `spacesize` function is now part of the public API, it also needs to be explicitly tested in the test suite. So please add a couple of simple tests and then merge :)

**Comment by AayushSabharwal** [comment]
> We can possibly extent `spacesize` to e.g., `OSMSpace` in a different PR.

The simplest method I can think of is looping over all nodes and finding the min/max lat/lon, and caching the result in the struct.

> @AayushSabharwal as this `spacesize` function is now part of the public API, it also needs to be explicitly tested in the test suite. So please add a couple of simple tests and then merge :)

Sure

**Comment by AayushSabharwal** [comment]
Will merge once tests pass. They work locally, and should pass here too but I'd rather make sure.

File: src/spaces/utilities.jl [hunk 7]
```diff
@@ -176,46 +202,21 @@ Example usage in [Battle Royale](
 function walk!(
     agent::AbstractAgent,
     direction::NTuple{D,Int},
-    model::ABM{<:AbstractGridSpace{D,true}};
-    kwargs...,
+    model::ABM{<:AbstractGridSpace};
+    ifempty::Bool = true
```
[the remark this answers is not in the document]
Comment (Datseris) [reply src/spaces/utilities.jl:206]: It's unfortunate that we didn't catch this. But please feel free to make a PR that fixes the Battle Royale example. Although I'm not sure what you mean by "it breaks the example in the docs", as the example runs fine here: https://juliadynamics.github.io/AgentsExampleZoo.jl/dev/examples/battle/

**Pull request merged**